### PR TITLE
Build: Run unit tests on more node versions, mac, and windows

### PR DESCRIPTION
**Pull request by IanVS**
Issue: N/A

## What I did

In https://github.com/storybookjs/storybook/pull/16723#pullrequestreview-811994587, it was requested to run tests on Windows.  It surprised me a bit that unit tests were not already automatically running on windows for every PR.  So, this PR adds windows to the unit tests in the github action.  I also took things a step further and added more versions of node.js, as well as MacOS.  

According to https://github.com/storybookjs/storybook/issues/16664, we can expect the windows tests to fail.  And I'm not sure if there's been discussion previously on adding windows tests or not, but I figured I'd throw this up and see what happens.  :)

## How to test

- [X] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by IanVS** [comment]
As expected, Windows tests are failing.  I'm not sure how we'd even get them passing, TBH.  Interestingly, Node 17 is also failing, due to something in webpack 5, it seems:

```
  this[kHandle] = new _Hash(algorithm, xofLen);
                  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:67:19)
    at Proxy.createHash (node:crypto:130:10)
    at module.exports (/home/runner/work/storybook/storybook/node_modules/webpack/lib/util/createHash.js:135:53)
    at NormalModule._initBuildHash (/home/runner/work/storybook/storybook/node_modules/webpack/lib/NormalModule.js:417:16)
    at handleParseError (/home/runner/work/storybook/storybook/node_modules/webpack/lib/NormalModule.js:471:10)
    at /home/runner/work/storybook/storybook/node_modules/webpack/lib/NormalModule.js:503:5
    at /home/runner/work/storybook/storybook/node_modules/webpack/lib/NormalModule.js:358:12
    at /home/runner/work/storybook/storybook/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/home/runner/work/storybook/storybook/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at Array.<anonymous> (/home/runner/work/storybook/storybook/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
    at Storage.finished (/home/runner/work/storybook/storybook/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
    at /home/runner/work/storybook/storybook/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
    at /home/runner/work/storybook/storybook/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (node:internal/fs/read_file_context:68:3) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v17.1.0
```

**Comment by IanVS** [comment]
One approach might be to use https://www.npmjs.com/package/jest-os-detection to write linux and windows specific tests that only run on particular operating systems.

**Comment by ndelangen** [comment]
Looks like (at least part of) the reason why tests are failing on windows is because ofline-ending issues:

Screenshot 2021-11-22 at 17 12 56

**Comment by IanVS** [comment]
@ndelangen exactly.  Line endings and path separator differences (`/` vs `\`) account for many of the failures, I believe.

**Comment by IanVS** [comment]
OK, it'll take a while for the CI to run, but I _think_ I've at least gotten the snapshots passing on both windows and linux.  I'll take a look at the unit tests next.

**Comment by IanVS** [comment]
@ndelangen is it possible that this is a legit failure on windows?

```
loadPreset › should prepend framework field to list of presets

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `loadPreset should prepend framework field to list of presets 1`

    - Snapshot  - 5
    + Received  + 0

    @@ -8,15 +8,10 @@
          "name": "@storybook/preset-typescript",
          "options": Object {},
          "preset": Object {},
        },
        Object {
    -     "name": "@storybook/addon-docs/preset",
    -     "options": Object {},
    -     "preset": Object {},
    -   },
    -   Object {
          "name": Object {
            "addons": Array [
              "@storybook/addon-docs",
            ],
            "framework": "@storybook/react",

      436 |       {}
      437 |     );
    > 438 |     expect(loaded).toMatchInlineSnapshot(`
          |                    ^
      439 |       Array [
      440 |         Object {
      441 |           "name": "@storybook/react",

      at Object.<anonymous> (lib/core-common/src/presets.test.ts:438:20)
```

**Comment by ndelangen** [comment]
That doesn't look good indeed. @IanVS

**Comment by Marklb** [comment]
I have taken a look at the Angular tests on Windows and usually end up just running the specific tests that I know my change could affect and just let the rest run in CI, or a linux VM if I am fairly sure I will need to rerun the tests multiple times. It would be nice to not have to do that. I am not the most experienced with unit tests, but I can try to help get the Angular ones passing, at least.

One tricky thing I have run into with tests on Windows is the different valid path formats, depending on what will be processing them. I was trying to fix one last night and I think it was a Typescript Webpack plugin had a path in the format `E:\Users\Docs\Project` then an Angular Webpack plugin had a path in the format `E:\\Users\\Docs\\Project` then something else had a path in the format `E:/Users/Docs/Project`.

**Comment by Marklb** [comment]
Well, I took a look at the Angular errors, again and even though I would like it to be more obvious which path formatting should be used, they weren't as far off from fixing as I remembered from last time I looked. I haven't tested them outside of windows yet, but I can do that then submit them or you can just pick them out of my fork if it's useful. It isn't much: https://github.com/Marklb/storybook/commit/e432ea09411439df37ef988b026722522f5ffae8

**Comment by IanVS** [comment]
Thanks for the help, @Marklb!  I've taken your approach and adjusted it slightly, using path.join instead of path.resolve, and using the `getSystemPath` utility from angular-devkit, although that may not have been strictly necessary.  I'm still not completely clear on why the two `.replace` are required, but I think one has to do with how jest snapshots are serialized.

**Comment by IanVS** [comment]
This looks like another potentially legitimate failure on windows: 

```
 eslint-plugin fix › when project does not contain eslint-plugin-storybook but has eslint installed › should no-op and warn when › .eslintrc is not found

    expect(jest.fn()).toHaveBeenCalledWith(...expected)

    Expected: "Unable to find .eslintrc config file, skipping"
    Received: "Unable to find storybook main.js config, skipping"

    Number of calls: 1

      90 |         });
      91 |
    > 92 |         expect(loggerSpy).toHaveBeenCalledWith('Unable to find .eslintrc config file, skipping');
         |                           ^
      93 |
      94 |         await expect(result).toBeFalsy();
      95 |       });

```

**Comment by Marklb** [comment]
@IanVS Using `getSystemPath` is what I initially tried, but kept have trouble with it. I probably tried it before realizing the snapshots were apparently escaping the `\`, at least I assume that is why the `replace` was needed.

I took a look at `lib\cli\src\automigrate\fixes\eslint-plugin.test.ts` and `lib\cli\src\automigrate\fixes\cra5.test.ts` and I got them to pass, but I don't think it is the right. In `__mocks__\fs-extra.js` I did a replace of all `\\` to `/` for each path passed into `__setMockFiles()` then replaced `\\` with `/` for each path doing a lookup in `mockFiles`.

At first I was thinking that a better solution would be to change the paths to the system path in `__setMockFiles()`, but Windows is fine with `\\` or `/` path separators. So, I think the methods would have to do some processing of the input path. If the methods would have to do some type of processing of the input path anyway then maybe my solution would work, but just validate that the input path is a valid system path first. Mapping them to a common format would probably be fine, as long as the methods still know when they receive a path that shouldn't be valid on that system.

The following is my `__mocks__\fs-extra.js`:
```js
const fs = jest.createMockFromModule('fs-extra');

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = Object.create(null);

// eslint-disable-next-line no-underscore-dangle
function __setMockFiles(newMockFiles) {
  mockFiles = newMockFiles;
  if (newMockFiles === undefined || newMockFiles === null) {
    mockFiles = null;
  }
  mockFiles = {};
  Object.keys(newMockFiles).forEach(k => {
    mockFiles[k.replace(/\\/g, '/')] = newMockFiles[k];
  });
}

// A custom version of `readdirSync` that reads from the special mocked out
// file list set via __setMockFiles
const readFile = async (filePath) => mockFiles[filePath.replace(/\\/g, '/')];
const readFileSync = (filePath = '') => mockFiles[filePath.replace(/\\/g, '/')];
const existsSync = (filePath) => !!mockFiles[filePath.replace(/\\/g, '/')];
const lstatSync = (filePath) => ({
  isFile: () => !!mockFiles[filePath.replace(/\\/g, '/')],
});

// eslint-disable-next-line no-underscore-dangle
fs.__setMockFiles = __setMockFiles;
fs.readFile = readFile;
fs.readFileSync = readFileSync;
fs.existsSync = existsSync;
fs.lstatSync = lstatSync;

module.exports = fs;
```

**Comment by IanVS** [comment]
Thanks yet again for the investigation, @Marklb.  After you pointed out what was going on, I realized that the calls to `__setMockFiles` was using a posix path instead of a windows path, and I think that changing those to a cross-platform compatible `path.join` fixes those issues.

**Comment by IanVS** [comment]
OK, I may need help with the rest of this.  There are only a few failures left:

```
Tests:       4 failed, 5 skipped, 2228 passed, 2237 total
Snapshots:   1 failed, 660 passed, 661 total
```

Several of them are related to `addon-docs`, and one is a timeout.

**Comment by IanVS** [comment]
Well, my muscle memory for copy and paste is completely screwed up now, but 🤞 that the tests should all pass (they do locally, at least).  I did need to change a bit of logic in a few places where I think the previous behavior in windows was incorrect, but it would be good if someone could take a look at those especially.  I'll call them out in PR comments after I confirm that CI does indeed pass.

**Comment by Marklb** [comment]
When the VSCode Jest extension `orta.vscode-jest` runs the test the following two files are created. I don't know much about Jest snapshots, but are those missing snapshots that should be tracked by git?
`lib\core-server\src\__snapshots__\cra-ts-essentials_manager-prod-windows`
`lib\core-server\src\__snapshots__\cra-ts-essentials_preview-prod-windows`

Also, the test in `addons\docs\src\frameworks\angular\angular-properties.test.ts` only passes if I use the `addons\docs\src\frameworks\angular\__testfixtures__\doc-button\compodoc-posix.snapshot`. I will see if I can tell where Compodoc gets the hash it appends to the id's, but I don't think the OS is what affects that.

**Comment by IanVS** [comment]
@Marklb that's strange, the windows tests passed on github CI just now, so I'm not sure what you're seeing.  I restarted them because I think a flakey test failed on mac.  

>are those missing snapshots that should be tracked by git?

I don't _think_ so, or else CI would fail, I'm pretty sure.  Try restarting vscode if you haven't lately (I had trouble with the jest extension creating files before I restarted it), and make sure that you've done a `yarn install` and rebuilt storybook recently, I guess.  I'm done on this for now (it's late here), but let me know if you make any discoveries.  Thanks!

**Comment by Marklb** [comment]
@IanVS Once I looked into how Compodoc generates the hash I realized the problem. That hash is generated as a sha512 hash of the source. By default, Git on Windows has `core.autocrlf` set to `true`. I always set that to `false`, because I have run into too many problems where linters, pre-commit hooks, or tests expect a specific line ending. *In my opinion it seems odd that the default would be to convert line endings when pulling and pushing, but that is apparently the default.*

In my repos I normally try to remember to add something like the following to a `.gitattributes` file, because that takes precedence over the `core.autocrlf` config value. I don't know what the best solution would be here. The options I see are to use `.gitattributes`, check the line endings or convert the line endings when reading the file.

**Comment by IanVS** [comment]
Thanks for digging into that @Marklb.  I don't know much about windows development, so I'll leave it to you and the other storybook maintainers to decide what to do. 

Speaking of which, I think this is ready for some eyes from folks like @ndelangen and @tmeasday.  I'd love if we could merge it fairly quickly, before other work happens which requires rebases or other failing windows tests are written.

File: .eslintrc.js
Comment (IanVS) [inline .eslintrc.js:8]: This gives us a way to write tests that don't run on windows or only run on windows.  There were only a handful of times that I needed this, and maybe there are other ways around it, but it seemed like a useful enough utility that I thought it would be worth adding.

File: addons/a11y/src/components/A11YPanel.test.tsx
Comment (IanVS) [inline addons/a11y/src/components/A11YPanel.test.tsx:139]: This timeout didn't actually seem to help, this test still fails intermittently.  Would love if someone who knows more about the a11y panel could take a look at why.  But even so, I think it might be good to merge this even knowing this test is flakey, just so we can get some windows tests running.

File: jest.config.js
Comment (IanVS) [inline jest.config.js:86]: By adding this boolean as a testing global, it makes it easier to consistently differentiate the os in our tests, mostly useful for snapshots, when combined with `toMatchSpecificSnapshot`.

File: lib/core-common/src/presets.ts
Comment (IanVS) [inline lib/core-common/src/presets.ts:90]: 🚨 This is a change to a non-test file.  The call to `resolveFrom` is expecting to resolve a node module, which always use forward slashes (it helped me to think about calls to `require()`).  Without this change, the mock of `resolve-from` in the tests would get windows paths, which it didn't know how to handle.  Also, I think that this is more-correct from a non-testing standpoint as well.

File: lib/core-common/src/utils/__tests__/paths.test.ts
Comment (IanVS) [inline lib/core-common/src/utils/__tests__/paths.test.ts:54]: I removed these because they were written in order to test windows-style paths on non-windows github machines.  Now that we can use `path` in the tests to actually create windows paths on windows, these duplicates are no longer needed.

File: lib/core-server/src/core-presets.test.ts
Comment (IanVS) [inline lib/core-server/src/core-presets.test.ts:150]: This test appeared to be quite flakey on mac and windows, seemingly due to tests stepping on each other, even though they run in sequence.  The changes I've made here do two main things:

- use nested `.each` blocks so that we can DRY it up a bit
- use a temporary folder to build storybook into, rather than building into the same directory in the repo each time and relying on it being cleaned up properly between tests.  I believe that is what was causing the flake, because after I changed to use a temp folder, those problems went away.

File: lib/core-server/src/core-presets.test.ts
Comment (IanVS) [inline lib/core-server/src/core-presets.test.ts:183]: Cache clearing is async, so we need to await it.

File: lib/core-server/src/utils/StoryIndexGenerator.ts
Comment (IanVS) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:181]: 🚨 This is another change to a non-test file.

File: lib/core-server/src/utils/watch-story-specifiers.ts
Comment (IanVS) [inline lib/core-server/src/utils/watch-story-specifiers.ts:83]: 🚨  Non-test file.  The documentation for `glob` states that it requires forward slashes, not windows paths, so I think this might have been causing problems on windows, although I can't find any issues reporting it.

**Comment by IanVS** [comment]
I talked to the team, and I'm going to set this up to only run on merges to main or if a new `test:node-os-matrix` (or something like that) label is used, now that we've seen that the tests pass here on this PR.  That's because testing takes such a long time, and blocks other tests from running in the queue.  

Another thing that might be worth looking into is speeding up the build time, since that's what is taking most of the time in the tests.  I think part of the issue there is that even though they're called `unit tests`, some of them are actually integration tests.  Otherwise we wouldn't need to build the entire monorepo before running them.

**Comment by IanVS** [comment]
OK, I've made some updates and verified that the unit test matrix jobs do not run on a PR unless the `ci:matrix` label is added, and that they start as soon as the label is added.  The tests should also run on merges to `next`.  I've just rebased on next, and believe this is ready for review now.

**Comment by IanVS** [comment]
OK, with the latest commit I've made sure the tests continue to run when new commits are pushed as long as that label still is present.

File: lib/core-server/src/core-presets.test.ts
Comment (kylegach) [inline lib/core-server/src/core-presets.test.ts:31]: Comments on this FIXME, @IanVS?

**Review by kylegach — COMMENTED** [review]
Thank you for this very thoughtful PR, @IanVS. Future contributors on Windows should have a much nicer experience.

These changes look good to me, but people more familiar with Windows dev and/or @shilman, @ndelangen, @tmeasday should provide final approval.

File: lib/core-server/src/core-presets.test.ts
Comment (IanVS) [reply lib/core-server/src/core-presets.test.ts:31]: I found that the `stories.json` file was still being created.  Also, when I turned on logging, I was seeing messages about this preset being invalid.  I think maybe it's not in the correct format for a preset, perhaps?

**Review by shilman — APPROVED** [review]
@IanVS This is amazing work. Thanks so much for putting the time in, and also for your patience on the merge. I'm guessing we'll need to tweak a bit once we run this in practice, but I'll coordinate that with you when the time comes. In the meantime, merging! 💯

**Comment by IanVS** [comment]
Thanks @shilman.  And just to close the loop on one point, it does look like the unit tests are running against the next branch, as intended:

image

**Comment by ndelangen** [comment]
I think running the unit tests on multiple platforms is great, but taking different snapshot on different platforms, will make things more difficult.

If someone needs to update the snapshots, they need access to both platforms, which is not something everyone has (me included, I do not have access to any windows platform device)

**Comment by IanVS** [comment]
That's a good point.  I was using a virtual machine on my mac when I was working on this PR, but that's a pain.  I could see a few different possible solutions.

1. Find a way to accept snapshot changes within CI (maybe this is possible?).
2. Let them fail until someone with windows can update them.
3. Remove the snapshots, replace with more explicit tests.
4. Remove windows tests in CI.

I'd love if we could find an option that's not #4, since I think it's good to have visibility into the windows experience.  But, I'll understand if that's the way y'all want to go.

**Comment by tmeasday** [comment]
> Find a way to accept snapshot changes within CI (maybe this is possible?).

I'm not necessarily advocating for this but with Circle it is possible to create downloadable artifacts from a build. What we did in a similar situation [in the Chromatic monorepo](https://github.com/chromaui/chromatic/blob/10ab7f9c3dfdb0796aa739b015b1d51e2bfed4e7/.circleci/config.yml#L229-L237) is do the snapshot tests in CI in two phases:

1. Run, record exit code.
2. If 1. failed, run a second time with `--updateSnapshot --onlyFailures`, place snapshots in artifact location
3. Return exit code from 1.

It's a bit clunky but it does mean as a dev you can download the snapshots generated in CI in 2.

**Pull request merged**
